src/hotspot/share/services/mallocHeader.hpp

```diff
@@ -140,10 +143,15 @@ class MallocHeader {
   // an option pointer to the corruption in p_corruption, and return false.
   // Return true if block is fine.
   inline bool check_block_integrity(char* msg, size_t msglen, address* p_corruption) const;
+  // Check correct alignment and placement of pointer, fill in short descriptive text and return false
+  // if this is not the case.
+  // Returns true if the memblock looks OK.
+  inline static bool is_valid_malloced_pointer(const void* payload, char* msg, size_t msglen);
 
   // If block is broken, print out a report to tty (optionally with
   // hex dump surrounding the broken block), then trigger a fatal error
-  inline void assert_block_integrity() const;
+  inline static const MallocHeader* resolve_checked(const void* memblock);
+  inline static MallocHeader* resolve_checked(void* memblock);
```

Comment:
Is there's a particular reason why in&out types need to match in terms of `const`? Does this work?

```cpp
  inline static MallocHeader* resolve_checked(const void* memblock);
```

Reply:
Good question. Perhaps in this specific case we can throw away the const information, but it's difficult to discover whether that's true and doesn't look correct to a reader at a first glance. I think an extra method reduces reading complexity such that it is worth keeping.

Reply:
You need both because you should be able to resolve a MallocHeader from a const payload pointer, but that should yield a const pointer too, not a non-const one. Yes, you could return a non-const header pointer, but that would be just bad const discipline.

Reply:
The input arg should be `const void*`, which accepts both `void*` and `const void*`, because that's a promise to the caller that the current method will not mutate the mem pointed by the input arg. (Ofc, one could throw away `const` and mutate the mem anyway inside the method, but that's just bad code.)

> but that should yield a const pointer too, not a non-const one.

I believe your reasoning is that in&out pointers are essentially the same (off by 1), so they should have the same const-ness.

However, the in&out args are two semantically different entities, `void*` vs `MallocHeader*`. For instance, a caller may construct a `MallocHeader*` out of a `const void*` -- the mem is immutable via `void*` but mutable via `MallocHeader*` due to the extra info offered by the more concrete type. IOW, whether the mem indicated by `MallocHeader*` is mutable or not is up to the caller.

Looking at the `const`-variant caller (I can just find this single one), it's more or less the same as a caller of the non-`const` variant.

```cpp
static void check_expected_malloc_header(const void* payload, MEMFLAGS type, size_t size) {
  const MallocHeader* hdr = MallocHeader::resolve_checked(payload);
  EXPECT_EQ(hdr->size(), size);
  EXPECT_EQ(hdr->flags(), type);
}
```

vs

```cpp
    void* const memblock = ...

    MallocHeader* header2 = MallocHeader::resolve_checked(memblock);
    assert(header2->size() == size, "Wrong size");
    assert(header2->flags() == flags, "Wrong flags");
```

(Seems to me, `header2` should be `const` as well -- here we promise not to mutate the obj via `header2`.)

Reply:
> The input arg should be `const void*`, which accepts both `void*` and `const void*`, because that's a promise to the caller that the current method will not mutate the mem pointed by the input arg. 

Yes,

(Ofc, one could throw away `const` and mutate the mem anyway inside the method, but that's just bad code.)

Yes.

> 
> > but that should yield a const pointer too, not a non-const one.
> 
> I believe your reasoning is that in&out pointers are essentially the same (off by 1), so they should have the same const-ness.
> 
> However, the in&out args are two semantically different entities, `void*` vs `MallocHeader*`. For instance, a caller may construct a `MallocHeader*` out of a `const void*` -- the mem is immutable via `void*` but mutable via `MallocHeader*` due to the extra info offered by the more concrete type. IOW, whether the mem indicated by `MallocHeader*` is mutable or not is up to the caller.

Not sure if we talk at cross purposes, but:

- const variant is for safe *querying* while preventing accidental write access. You don't need write access for querying.
- non-const variant is for modifying the header. The header is mostly immutable, the only thing we change are the canaries. Without that, we could dispose of the non-const variant altogether.

The whole thing follows the same logic as a class where you differentiate between read-only and write-access for the same accessor, e.g.

```
class X {
   Y y;
   const Y& get_y() const;
   Y& get_y();
}
```

see e.g. Effective C++. Regrettably, C++ forces us to duplicate the method.

> 
> Looking at the `const`-variant caller (I can just find this single one), it's more or less the same as a caller of the non-`const` variant.
> 
> ```c++
> static void check_expected_malloc_header(const void* payload, MEMFLAGS type, size_t size) {
>   const MallocHeader* hdr = MallocHeader::resolve_checked(payload);
>   EXPECT_EQ(hdr->size(), size);
>   EXPECT_EQ(hdr->flags(), type);
> }
> ```
> 
> vs
> 
> ```c++
>     void* const memblock = ...
> 
>     MallocHeader* header2 = MallocHeader::resolve_checked(memblock);
>     assert(header2->size() == size, "Wrong size");
>     assert(header2->flags() == flags, "Wrong flags");
> ```
> 
> (Seems to me, `header2` should be `const` as well -- here we promise not to mutate the obj via `header2`.)

Agreed.

Reply:
When you agree with "The input arg should be `const void*`...", do you mean this?

```cpp
  inline static const MallocHeader* resolve_checked(const void* memblock);
  inline static       MallocHeader* resolve_checked(const void* memblock);
```
I don't think this is valid C++.

What I suggested is to have a single method and let the caller decide regarding `const` of the returned value.

Reply:
> When you agree with "The input arg should be `const void*`...", do you mean this?
> 
> ```c++
>   inline static const MallocHeader* resolve_checked(const void* memblock);
>   inline static       MallocHeader* resolve_checked(const void* memblock);
> ```
> 
> I don't think this is valid C++.
> 
> What I suggested is to have a single method and let the caller decide regarding `const` of the returned value.

No, I did not mean that. Both functions are needed. What you suggest would lead to a violation of constness. Why drop protection? To gain what, save one line?

Having two functions that differ in constness of input arg (or of this pointer) and output argument is an established practice. The normal recommendation for non-trivial functions then is to implement one in terms of the other by casting constness away. @jdksjolen solved this in a neater, albeit more verbose way with a templatized implementation.

Reply:
> Both functions are needed.

The signature of `inline static MallocHeader* resolve_checked(void* memblock);` is too permissive -- the callee is allowed to mutate content of `memblock`, but it shouldn't.

> What you suggest would lead to a violation of constness.

As I pointed out before, constructing a `MallocHeader*` from a `const void*` can be a valid use case.

> Why drop protection? To gain what, save one line?

I think the bigger issue is having `void*` as the arg, which forces callers to check the impl of this method.